internal/icinga2/client_api.go

```diff
@@ -256,46 +271,74 @@ func (client *Client) checkMissedChanges(ctx context.Context, objType string, ca
 			continue
 		}
 
-		// First: State change event
-		ev, err := client.buildHostServiceEvent(
-			ctx,
-			objQueriesResult.Attrs.LastCheckResult, objQueriesResult.Attrs.State,
-			hostName, serviceName)
+		attrs := objQueriesResult.Attrs
+
+		// First: State change event, so that an incident is created if there's none yet.
+		ev, err := client.buildHostServiceEvent(ctx, attrs.LastCheckResult, attrs.State, hostName, serviceName)
 		if err != nil {
 			return fmt.Errorf("failed to construct Event from Host/Service response, %w", err)
 		}
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		case catchupEventCh <- &catchupEventMsg{eventMsg: &eventMsg{ev, objQueriesResult.Attrs.LastStateChange.Time()}}:
+		case catchupEventCh <- &catchupEventMsg{eventMsg: &eventMsg{ev, attrs.LastStateChange.Time()}}:
 			stateChangeEvents++
 		}
 
-		// Second: Optional acknowledgement event
-		if objQueriesResult.Attrs.Acknowledgement == 0 {
+		// Skip if there's no active incident for this Checkable.
+		// StateServiceOk is 0 and so is StateHostUp, thus this applies to both of them.
+		if attrs.State == StateServiceOk {
+			client.Logger.Debugw("Skipping check for (un)mute events, checkable recovered",
+				zap.String("object_type", objType), zap.String("host_name", hostName),
+				zap.String("service_name", serviceName))
 			continue
 		}
 
-		ackComment, err := client.fetchAcknowledgementComment(
-			ctx,
-			hostName, serviceName,
-			objQueriesResult.Attrs.AcknowledgementLastChange.Time())
-		if err != nil {
-			return fmt.Errorf("fetching acknowledgement comment for %v failed, %w", ev, err)
-		}
+		var fakeEv *event.Event
+		if attrs.IsFlapping || attrs.DowntimeDepth > 0 {
+			fakeEv, err = client.buildCommonEvent(ctx, hostName, serviceName)
+			if err != nil {
+				return fmt.Errorf("failed to construct checkable fake mute event: %w", err)
+			}
 
-		ev, err = client.buildAcknowledgementEvent(
-			ctx,
-			hostName, serviceName,
-			ackComment.Author, ackComment.Text, false)
-		if err != nil {
-			return fmt.Errorf("failed to construct Event from Acknowledgement response, %w", err)
+			fakeEv.Type = event.TypeMute
+			if attrs.DowntimeDepth > 0 {
+				fakeEv.SetMute(true, "Checkable is in downtime, but we missed the Icinga 2 DowntimeStart event")
+			} else {
+				fakeEv.SetMute(true, "Checkable is flapping, but we missed the Icinga 2 FlappingStart event")
+			}
+		} else if attrs.Acknowledgement > AcknowledgementNone {
```

Comment:
> Also, I'd just use `!=` for the comparison. I don't see a need why we should have to rely on the exact underlying integer values.

I have absolutely no idea what you're talking about, since when does the `>` check rely on the exact integer values? It's clearly documented what values [`Acknowledgement`](https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/#event-stream-type-acknowledgementset) can have, so I don't really know what to say.

Reply:
From the functionality, it's exactly the same, but:

- `attrs.Acknowledgement != AcknowledgementNone` simply says "anything except no acknowledgement"
- `attrs.Acknowledgement > AcknowledgementNone` raises the question what's the value of `AcknowledgementNone` and how does it compare to other possible values, which you have to look up and in the end conclude "anything except no acknowledgement"